docarray/index/backends/mongodb_atlas.py

```diff
@@ -0,0 +1,519 @@
+import collections
+from collections import defaultdict
+from dataclasses import dataclass, field
+from functools import cached_property
+
+# from importlib.metadata import version
+from typing import (
+    Any,
+    Dict,
+    Generator,
+    Generic,
+    List,
+    Optional,
+    Sequence,
+    Type,
+    TypeVar,
+    Union,
+)
+
+import bson
+import numpy as np
+from pymongo import MongoClient
+
+from docarray import BaseDoc, DocList
+from docarray.index.abstract import BaseDocIndex, _raise_not_composable
+from docarray.typing.tensor.abstract_tensor import AbstractTensor
+from docarray.utils._internal._typing import safe_issubclass
+from docarray.utils.find import _FindResult, _FindResultBatched
+
+# from pymongo.driver_info import DriverInfo
+
+
+MAX_CANDIDATES = 10_000
+OVERSAMPLING_FACTOR = 10
+TSchema = TypeVar('TSchema', bound=BaseDoc)
+
+
+class MongoDBAtlasDocumentIndex(BaseDocIndex, Generic[TSchema]):
+    def __init__(self, db_config=None, **kwargs):
+        super().__init__(db_config=db_config, **kwargs)
+        self._create_indexes()
+        self._logger.info(f'{self.__class__.__name__} has been initialized')
+
+    @property
+    def _collection(self):
+        if self._is_subindex:
+            return self._db_config.index_name
+
+        if not self._schema:
+            raise ValueError(
+                'A MongoDBAtlasDocumentIndex must be typed with a Document type.'
+                'To do so, use the syntax: MongoDBAtlasDocumentIndex[DocumentType]'
+            )
+
+        return self._schema.__name__.lower()
+
+    @property
+    def index_name(self):
+        """Return the name of the index in the database."""
+        return self._collection
+
+    @property
+    def _database_name(self):
+        return self._db_config.database_name
+
+    @cached_property
+    def _client(self):
+        return self._connect_to_mongodb_atlas(
+            atlas_connection_uri=self._db_config.mongo_connection_uri
+        )
+
+    @property
+    def _doc_collection(self):
+        return self._client[self._database_name][self._collection]
+
+    @staticmethod
+    def _connect_to_mongodb_atlas(atlas_connection_uri: str):
+        """
+        Establish a connection to MongoDB Atlas.
+        """
+
+        client = MongoClient(
+            atlas_connection_uri,
+            # driver=DriverInfo(name="docarray", version=version("docarray"))
+        )
+        return client
+
+    def _create_indexes(self):
```

Comment:
this is not implemented?

Reply:
creating indexes programatically is not available on free tier currently, hence skipping it